packages/medusa/src/interfaces/transaction-base-service.ts

```diff
@@ -67,12 +76,22 @@ export abstract class TransactionBaseService {
       dontFail = !!maybeErrorHandlerOrDontFail
     }
 
+    // If the transaction manager is already set, we are in an ongoing
+    // transaction and therefore we should use that manager for subsequent work.
     if (this.transactionManager_) {
       const doWork = async (m: EntityManager): Promise<never | TResult> => {
         this.manager_ = m
         this.transactionManager_ = m
+
+        // If no transaction id is provided, we generate uuid to use
+        const txId = options.transactionId ?? this.transactionId_ ?? v4()
+        this.transactionId_ = txId
+
         try {
-          return await work(m)
+          const result = await work(m, txId)
+          // After the transaction is complete, we process cached events
+          this.eventBusService_.processCachedEvents(txId)
```

Comment:
thoughts: I was thinking, would it make sense to directly create something like
```ts
app.use(function(req, res, next) {
  req.eventTransactionId = v4()

  const eventBusService = req.resolve(...)

  res.on('close', function() {
    eventBusService_.processCachedEvents(txId)
  })
})
```

So eventually we don't have to touche the actual atomicPhase and instead either create the context object to pass to the method being called or use the transaction manager to only pass the id down to the handler

And we can invalidate the cache in the error handler.
It might be something for the future but I just wanted to propose it for future ideas 😂

Reply:
Do you think this is a better approach now given that we want to move towards service to service communication in the future i.e. not HTTP (express)?

Reply:
I was thinking that since the API will act as the entry point and termination, it could be managed that way. then when the service will communicate between them, as we said there will be a contextual object passed along in the message or whatever, but this proposal above would act similarly to the transaction orchestrator until it is finished. Basically, it is the start and end of the flow

Reply:
The main advantage is that it is not related to anything specifically related to the transaction as this pattern can be used more widely than the transaction it self as mention in a previous discussion. I think that it gives more flexibility in regards to the user caching the events even in his customisation or out of a transaction but part of a wider flow